src/sentry/receivers/outbox.py

```diff
@@ -0,0 +1,137 @@
+from __future__ import annotations
+
+import random
+import sys
+from typing import Any, Callable, List, Mapping
+
+from django.db.models.signals import post_save, pre_delete
+from django.dispatch import receiver
+
+from sentry import options
+from sentry.models import (
+    ControlOutbox,
+    Organization,
+    OrganizationMember,
+    OutboxBase,
+    OutboxCategory,
+    RegionOutbox,
+    User,
+    outbox_silo_modes,
+    process_region_outbox,
+)
+from sentry.services.hybrid_cloud.organization_mapping import (
+    ApiOrganizationMappingUpdate,
+    organization_mapping_service,
+)
+from sentry.services.hybrid_cloud.tombstone import ApiTombstone, tombstone_service
+from sentry.silo import SiloMode
+
+
+# This method is split out for test mocking convenience, and for future potential logging / metrics
+def _prepare_outbox(outbox: OutboxBase):
+    if random.random() < float(options.get("hybrid_cloud.outbox_rate")) or "pytest" in sys.modules:
+        outbox.save()
+
+
+def add_to_region_outbox(instance: Any):
+    _prepare_outbox(RegionOutbox.for_model_update(instance))
+
+
+def add_to_control_outbox(instance: Any):
+    for outbox in ControlOutbox.for_model_update(instance):
+        _prepare_outbox(outbox)
+
+
+@receiver(post_save)
+@receiver(pre_delete)
+def process_outbox(instance: Any, signal: Any, **kwargs: Any):
```

Comment:
I don't disagree, especially if this were a pattern that needed to be placed in other spots more, but I also don't agree especially either.

For one, annoyingly, adding any of this logic directly to model files introduces IMPORT HELL.  A ton of if TYPE_CHECKING and lazy imports have to carefully be aligned and adjusted everytime something changes due to the need to import the enums, the outboxes, etc, etc.  Keeping it in a separate file here ensures that as you add more outbox patterns, the import fight does not constantly increase in complexity.

For two, I don't forsee us using outbox more than strictly necessary.  If a developer's first thought is "I want to add some outboxing to my model", their first step should be to read code and find things.  I really don't want this system to appear magical -- I want it to be a thoughtful integration that only happens *when it is absolutely necessary and other options are not sufficient for a defensible reason*.

I'll consider generalizing or improving UX of the api as we continue to expand its integration however.

Reply:
I would also prefer to have a more explicit opt-in flow for future outbox additions. My thinking a few months back were that we would modify endpoint/application logic to explicitly push into the outbox. Signals work well, but they can create 'hidden' side-effects and make following application logic harder as the control flow jumps around more.

I think we could be well served by a dumb/simple solution like:

```python
with transaction.atomic():
   member.save()
   RegionOutbox.create_membership_update(member)
```

This approach would be simpler to build/operate/review. It would be more prone to mistakes and missed updates though. Signals adds more comprehensive capturing at the cost of more complex code and having to know where/how to wire the application together.

Reply:
I definitely dislike signals for many reasons.  You aren't going to get a comprehensive defense of it from me.

Better, I love *systems with strong guarantees that don't require exponential growth of required context knowledge to operate and update*.  For instance, *triggers* provide strong guarantees at a fundamental, and extremely performant level of the system.

The explicit system can work, there is no doubt.  But what's the workflow?   I need to replicate system deletes of Users.  How, will I know it is safe to remove the foreign key without dangling PII in our system?  Hard reconciliation is especially problematic with deletes, and will be much much more onerous and error prone when we actually have 2 or more regions.

Reply:
Arg, this problem drives me nuts. 

Basically, honestly, I think the only good and correct solution is triggers.  Everything else is gonna suck for a variety of annoying reasons.

Reply:
Actually, yeah, you may be right.  CDC can help create missing transaction *in the case that read after write isn't required*.  And in the case that it is required, there would be an explicit opt-in because it would be required for the test to pass (or atleast, we would need to design our tests that way).

Really, we need CDC in addition to this system.

Reply:
Yeah, but then the problem is that mixing asynchronous CDC reconciliation with explicit, synchronous flushing (for read after write, see org member creation) will lead to race conditions and ordering issues (flushing the org member before the org outbox is created by the async CDC process).  So that can't be the full solution either.

Reply:
The goal is that the order of the transactions that create or delete objects is preserved in the resulting outbox AND in such a way that a controller can synchronously flush.  In that state, we're really not left with many good options.